servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/RepeatWhenSingleTest.java

```diff
@@ -0,0 +1,272 @@
+
+    @Test
+    void repeaterThrows() {
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> {
+                    throw DELIBERATE_EXCEPTION;
+                }))
+                .subscribe(subscriber);
+        Subscription s = subscriber.awaitSubscription();
+        s.request(2);
+        assertThat(subscriber.takeOnNext(), equalTo(1));
+        assertThat(subscriber.awaitOnError(), is(DELIBERATE_EXCEPTION));
+    }
+
+    @Test
+    void invalidDemand() {
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount == 1 ? completed() : failed(new DeliberateException())))
+                .subscribe(subscriber);
+        subscriber.awaitSubscription().request(-1);
+        assertThat(subscriber.awaitOnError(), instanceOf(IllegalArgumentException.class));
+    }
+
+    @Test
+    void concurrentInvalidDemand() throws BrokenBarrierException, InterruptedException {
+        CyclicBarrier barrier = new CyclicBarrier(2);
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount == 1 ? executor.submit(() -> {
+                    try {
+                        barrier.await();
+                    } catch (Exception e) {
+                        throw new AssertionError(e);
+                    }
+                }) : failed(new DeliberateException())))
+                .subscribe(subscriber);
+        Subscription s = subscriber.awaitSubscription();
+        s.request(1);
+        barrier.await();
+        s.request(-1);
+        s.request(-1); // test multiple invalid request-n should only terminate one time.
+        assertThat(subscriber.takeOnNext(), equalTo(1));
+        assertThat(subscriber.awaitOnError(), instanceOf(IllegalArgumentException.class));
+    }
+
+    @Test
+    void invalidDemandAfterFirst() {
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount == 1 ? completed() : failed(new DeliberateException())))
+                .subscribe(subscriber);
+        Subscription s = subscriber.awaitSubscription();
+        s.request(1);
+        assertThat(subscriber.takeOnNext(), equalTo(1));
+        s.request(-1);
+        assertThat(subscriber.awaitOnError(), instanceOf(IllegalArgumentException.class));
+    }
+
+    @Test
+    void singleRepeatAllDemandUpfront() {
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount == 1 ? completed() : failed(new DeliberateException())))
```

Comment:
In case users will look at these tests as an example, consider using constant `DELIBERATE_EXCEPTION` instead. Allocationg a new exception is costly if they follow the same pattern.

Reply:
sounds good, just want to differentiate for tests that actually expect `DELIBERATE_EXCEPTION` to propagate, but that set of tests is smaller so I can reduce allocation.